### PR TITLE
update to calcium imaging demo

### PR DESCRIPTION
**Pull request by EricThomson**
This pertains to Issue #151, mostly minor update to the calcium imaging demo. I did my best to address many of the things I mentioned in that issue, but not all. For instance:

1. I said it was `GCaMP6f` I'm not sure if this is right can we find out? :smile: 
2. The Pynapple object has lots of data we don't discuss it might be helpful to very briefly mention. E.g., are `x,y,z` animal location? are `rx` and `rz` head direction like `ry`, but on a different plane?
3. Is there an advantage of the soft-plus nonlinearity when using calcium imaging data?
4. In general at the end it sort of just sputters out -- we show how to fit the GLM to the data, and compare to linear regression, but don't really discuss or compare and contrast. Also, what might be added to improve the fit, reduce the bias (the bias looks fairly high)? Tweaking the regularization? Am I right in assuming we only have activity history filter, but not inter-neuron filter? As it stands the results are pretty far off, so I think this needs to be addressed. I would probably low-pass filter the signals with a Gaussian (on top of the subsampling already done) and then re-run it to see how it does, it just seems to not be handling the high frequency stuff very well (but this is conjecture on my part).

### CODE REVIEW

**Comment by BalzaniEdoardo** [comment]
> This pertains to Issue #151, mostly minor update to the calcium imaging demo. I did my best to address many of the things I mentioned in that issue, but not all. For instance:
> 
>     1. I said it was `GCaMP6f` I'm not sure if this is right can we find out? 😄
> 
>     2. The Pynapple object has lots of data we don't discuss it might be helpful to very briefly mention. E.g., are `x,y,z` animal location? are `rx` and `rz` head direction like `ry`, but on a different plane?
> 
>     3. Is there an advantage of the soft-plus nonlinearity when using calcium imaging data?
> 
>     4. In general at the end it sort of just sputters out -- we show how to fit the GLM to the data, and compare to linear regression, but don't really discuss or compare and contrast. Also, what might be added to improve the fit, reduce the bias (the bias looks fairly high)? Tweaking the regularization? Am I right in assuming we only have activity history filter, but not inter-neuron filter? As it stands the results are pretty far off, so I think this needs to be addressed. I would probably low-pass filter the signals with a Gaussian (on top of the subsampling already done) and then re-run it to see how it does, it just seems to not be handling the high frequency stuff very well (but this is conjecture on my part).

Hi Eric, for the bias stand point, we can cross-validate the regularizer. 

```python
from sklearn.model_selection import GridSearchCV
# assume you have the glm model
grid = dict(regularizer__regularizer_strength=[0.001, 0.01, 0.1])
clf = GridSearchCV(model, grid)
clf.fit(X, Y)
```
This should work, modulo "regularizer__regularizer_strength" being the right parameter name. You can check the key to use,
```
print(model.get_params())
``` 

For what it concern what we are fitting, we are fitting only the coupling filters from other neuron, not the autocorr filter. If you add an autocorr filter, there is a long autocorr, that will result in overfitting

File: docs/neural_modeling/plot_06_calcium_imaging.py [hunk 1]
```diff
@@ -103,26 +126,26 @@
...
 basis = heading_basis + coupling_basis
 
+
 # %%
 # ## Gamma GLM
 #
-# Since the transients are non-negative, we will use a Gamma distribution from `nemos` with a soft-plus non linearity.
-#
+# Until now, we have been modeling spike trains, and have used a Poisson distribution for the observability model. With calcium traces, things are quite different: we no longer have counts but continuous signals, so the Poisson is no longer appropriate. We cannot use a Gaussian because the calcium traces are non-negative. To satisfy these constraints, we will use a Gamma distribution from `nemos` with a soft-plus non linearity.
+
```
Comment (BalzaniEdoardo) [inline docs/neural_modeling/plot_06_calcium_imaging.py:139]: Following up on your comment, we can add a note about the non-linearity. something of this kind
```suggestion
!!! note "Non-linearity"
    Different option are possible. With a soft-plus we are assuming an "additive" effect the predictors, while an exponential non-linearity assumes multiplicative effects. Deciding which firing rate model works best is an empirical question. You can fit different configurations to see which one capture best the neural activity.
```

File: docs/neural_modeling/plot_06_calcium_imaging.py [hunk 2]
```diff
@@ -103,26 +126,26 @@
...
 
 model = nmo.glm.GLM(
     regularizer=nmo.regularizer.UnRegularized(solver_name="LBFGS", solver_kwargs=dict(tol=10**-13)),
     observation_model=nmo.observation_models.GammaObservations(inverse_link_function=jax.nn.softplus)
 )
 
 # %%
-# We select a neuron to fit and remove it from the list of predictors
+# We select one neuron to fit later, so remove it from the list of predictors
```
Comment (BalzaniEdoardo) [inline docs/neural_modeling/plot_06_calcium_imaging.py:147]: In this case, we are removing the self-coupling from the predictors. If this wasn't clear to you. @EricThomson we should rephrase it. The neuron was left out because of the long auto-corr time; the autocorr filter resulted in over fitting.

File: docs/neural_modeling/plot_06_calcium_imaging.py [hunk 3]
```diff
@@ -215,6 +239,8 @@
 plt.show()
 
 # %%
+# While there is some variability in the fit for both models, one advantage of the gamma distribution is clear: the nonnegativity constraint is followed with the data. 
+#
```
Comment (BalzaniEdoardo) [inline docs/neural_modeling/plot_06_calcium_imaging.py:243]: Maybe we can add, 
```suggestion
# This is needed if you want to use GLMs for predict the firing rate, which needs to be positive, in response a simulated input, as shown in [this paper](https://www.nature.com/articles/s41467-017-01908-3) for a Poisson GLM.

**Review by BalzaniEdoardo — APPROVED** [review]
Thank you Eric, this is a very nice editing!

I added a couple of comments to this but looks good to me!

File: docs/neural_modeling/plot_06_calcium_imaging.py (hunk 1 above)
Comment (EricThomson) [reply docs/neural_modeling/plot_06_calcium_imaging.py:139]: These are all helpful feedback, I will incorporate asap!

**Comment by gviejo** [comment]
Yes it's gcamp6f

**Comment by EricThomson** [comment]
Thanks a lot Edoardo for this  reply. I'm adjusting things accordingly. I tried running on some Gaussian smoothed data -- it did not go well, things looked ... not better :smile: I'm adding a regularizer so the tuning curves seem to better match the actual, and also this might give students a chance to tweak it (I did a manual grid search :upside_down_face: ).

**Comment by EricThomson** [comment]
I added a little thing at the end about how this is a more experimental tool than the stuff for spiking models. Feel free to alter/cut I often err on the side of verbose or overly flowery language. :rose: 

I could do a lot more with this, but I'm about out of time I have to get ready for a trip soon so have at it; what you did looks really good, I just made some minor adjustments! :fire:

File: docs/neural_modeling/plot_06_calcium_imaging.py [hunk 4]
```diff
@@ -91,38 +114,43 @@
...
+heading_basis.identifiability_constraints = True
 basis = heading_basis + coupling_basis
 
+
 # %%
 # ## Gamma GLM
 #
-# Since the transients are non-negative, we will use a Gamma distribution from `nemos` with a soft-plus non linearity.
-#
+# Until now, we have been modeling spike trains, and have used a Poisson distribution for the observability model. With calcium traces, things are quite different: we no longer have counts but continuous signals, so the Poisson is no longer appropriate. We cannot use a Gaussian because the calcium traces are non-negative. To satisfy these constraints, we will use a Gamma distribution from `nemos` with a soft-plus non linearity.
```
Comment (BalzaniEdoardo) [inline docs/neural_modeling/plot_06_calcium_imaging.py]:
```suggestion
# Until now, we have been modeling spike trains, and have used a Poisson distribution for the observability model. With calcium traces, things are quite different: we no longer have counts but continuous signals, so the Poisson is no longer appropriate. A a Gaussian model is not ideal since the calcium traces are non-negative. To satisfy these constraints, we will use a Gamma distribution from NeMoS with a soft-plus non linearity.
```

**Review by BalzaniEdoardo — APPROVED** [review]
I'll add a couple of commits, but this is great. the fit looks better too
